R/utilities-simulation.R

```diff
@@ -749,3 +749,230 @@ getSimulationTree <- function(simulationOrFilePath, quantityType = "Quantity") {
+#' Default value is 1e-15.
+#' @param simulationRunOptions Optional instance of a `SimulationRunOptions`
+#'  used during the simulation run.
+#'
+#' @return A named list, where the names are the IDs of the simulations and the
+#'   entries are lists containing `paths` and their `values` at the end of the
+#'   simulation.
+#' @import ospsuite.utils
+#' @export
+#' @examples
+#' simPath <- system.file("extdata", "Aciclovir.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#' steadyState <- getSteadyState(simulations = sim)
+#' # Set initial values for steady-state simulations
+#' setQuantityValuesByPath(
+#'   quantityPaths = steadyState[[sim$id]]$paths,
+#'   values = steadyState[[sim$id]]$values, simulation = sim
+#' )
+getSteadyState <- function(quantitiesPaths = NULL,
+                           simulations,
+                           steadyStateTime = 1000,
+                           ignoreIfFormula = TRUE,
+                           stopIfNotFound = TRUE,
+                           lowerThreshold = 1e-15,
+                           simulationRunOptions = NULL) {
+  ospsuite.utils::validateIsOfType(simulations, type = "Simulation")
+  ospsuite.utils::validateIsString(quantitiesPaths, nullAllowed = TRUE)
+  ospsuite.utils::validateIsNumeric(steadyStateTime, nullAllowed = FALSE)
+  simulations <- ospsuite.utils::toList(simulations)
+
+  if (any(steadyStateTime <= 0)) {
+    stop(messages$steadyStateTimeNotPositive(steadyStateTime))
+  }
+
+  # If `steadyStateTime` is a vector of values, it must be of the same size as
+  # the list of simulations.
+  # Otherwise, repeat the value for the number of simulations
+  if (length(steadyStateTime) > 1) {
+    ospsuite.utils::validateIsSameLength(simulations, steadyStateTime)
+  } else {
+    steadyStateTime <- rep(steadyStateTime, length(simulations))
+  }
+
+  # First prepare all simulations by setting their outputs and time intervals
+  # If no quantities have been specified, the quantities paths may be different
+  # for each simulation and must be stored separately
+  simulationState <- .storeSimulationState(simulations)
+  quantitiesPathsMap <- vector(mode = "list", length = length(simulations))
+  for (idx in seq_along(simulations)) {
+    simulation <- simulations[[idx]]
+    simId <- simulation$id
+    # Set simulation time to the steady-state value.
+    ospsuite::clearOutputIntervals(simulation = simulation)
+    simulation$outputSchema$addTimePoints(timePoints = steadyStateTime[[idx]])
+    # If no quantities are explicitly specified, simulate all outputs.
+    if (is.null(quantitiesPaths)) {
+      quantitiesPathsMap[[idx]] <- ospsuite::getAllStateVariablesPaths(simulation)
+    } else {
+      quantitiesPathsMap[[idx]] <- quantitiesPaths
+    }
+    names(quantitiesPathsMap)[[idx]] <- simId
+    ospsuite::clearOutputs(simulation)
+    ospsuite::addOutputs(quantitiesOrPaths = quantitiesPathsMap[[idx]], simulation = simulation)
```

Comment:
It would be nice to have the function `setOutputs(quantityPaths)` in the ospsuite-R, which would just execute the 2 functions above (clearOutputs and addOutputs )

Reply:
PR created